crates/bevy_ecs/src/system/input.rs

```diff
@@ -222,9 +222,9 @@ impl<'i, T: ?Sized> DerefMut for InMut<'i, T> {
 /// Used for [`ObserverSystem`]s.
 ///
 /// [`ObserverSystem`]: crate::system::ObserverSystem
-impl<E: 'static, B: Bundle> SystemInput for On<'_, E, B> {
-    type Param<'i> = On<'i, E, B>;
-    type Inner<'i> = On<'i, E, B>;
+impl<E: Event, B: Bundle> SystemInput for On<'_, '_, E, B> {
+    type Param<'i> = On<'i, 'i, E, B>;
+    type Inner<'i> = On<'i, 'i, E, B>;
```

Comment:
I would add a note here that this is a key detail for the safety of `On` and the fact we must use one lifetime here is the source of the issues we have in other places.